storage/src/main/java/org/apache/kafka/storage/internals/log/FetchDataInfo.java

```diff
@@ -39,19 +40,40 @@ public FetchDataInfo(LogOffsetMetadata fetchOffsetMetadata,
                          Records records,
                          boolean firstEntryIncomplete,
                          Optional<List<FetchResponseData.AbortedTransaction>> abortedTransactions) {
-        this(fetchOffsetMetadata, records, firstEntryIncomplete, abortedTransactions, Optional.empty());
+        this(fetchOffsetMetadata,
+             records,
+             firstEntryIncomplete,
+             abortedTransactions,
+             Optional.empty(),
+             LogOffsetMetadata.UNKNOWN_OFFSET_METADATA);
     }
 
     public FetchDataInfo(LogOffsetMetadata fetchOffsetMetadata,
                          Records records,
                          boolean firstEntryIncomplete,
                          Optional<List<FetchResponseData.AbortedTransaction>> abortedTransactions,
-                         Optional<RemoteStorageFetchInfo> delayedRemoteStorageFetch) {
+                         Optional<RemoteStorageFetchInfo> delayedRemoteStorageFetch,
+                         LogOffsetMetadata maxOffsetMetadata) {
         this.fetchOffsetMetadata = fetchOffsetMetadata;
         this.records = records;
         this.firstEntryIncomplete = firstEntryIncomplete;
         this.abortedTransactions = abortedTransactions;
         this.delayedRemoteStorageFetch = delayedRemoteStorageFetch;
+        this.maxOffsetMetadata = maxOffsetMetadata;
+    }
+
+    public FetchDataInfo withMaxOffsetMetadata(LogOffsetMetadata maxOffsetMetadata) {
+        return new FetchDataInfo(fetchOffsetMetadata,
+                                 records,
+                                 firstEntryIncomplete,
+                                 abortedTransactions,
+                                 delayedRemoteStorageFetch,
+                                 maxOffsetMetadata);
+    }
+
+    public boolean isLastSegment() {
```

Comment:
last segment is always the active segment. Shall we call it isActiveSegment()? 

Also, FetchDataInfo is not a property of a segment, hence isActiveSegment() sounds counter intuitive. But it also cannot have data spanning across multiple segments. So, perhaps rename the function to, isFetchDataFromActiveSegment()

Reply:
Please add documentation and state and there may be cases where we don't know (unknown_offset_metatda) and in such cases we return false, hence, this function may return falseNegative.